File: java/188_BestTimeToBuyAndSellStockIV.java
class Solution {
    public int maxProfit(int k, int[] prices) {
        int N = prices.length;
        if (k >= N/2) {
            int profit = 0;
            for (int i = 1; i < prices.length; ++i) profit += Math.max(prices[i] - prices[i-1], 0);
            return profit;
        }
        
        int[][] dp = new int[k+1][N];
        for (int i = 1; i <= k; ++i) {
            int profit = -prices[0];
            for (int j = 1; j < N; ++j) {
                dp[i][j] = Math.max(dp[i][j-1], profit + prices[j]);
                profit = Math.max(dp[i-1][j] - prices[j], profit);
            }
        }
        return dp[k][N-1];
    }
}

File: python/188_BestTimeToBuyAndSellStockIV.py
class Solution:
    def maxProfit(self, k: int, prices: List[int]) -> int:
        N = len(prices)
        if 2*k >= N:
            profit = 0
            for i in range(1, N):
                profit += max(0, prices[i] - prices[i-1])
            return profit
        
        dp = [[0 for j in range(N)] for i in range(k+1)]
        for i in range(1, k+1):
            max_profit = -prices[0]
            for j in range(1, N):
                dp[i][j] = max(dp[i][j-1], max_profit + prices[j])
                max_profit = max(dp[i-1][j] - prices[j], max_profit)
        
        return dp[k][N-1]
